### DISCUSSION TITLE
running the application at the development stage directly on Android using USB

### DISCUSSION THREAD

**Discussion by nexblu** [flet-dev/flet#3046] [Ideas]
excuse me sir, i want ask, Can flet be live debugged using a USB plugged directly into an Android device?

**Comment by ndonkoHenri** [comment]
I personally don't really seem to understand the question. Can you better explain what you mean?

**Reply by nexblu** [reply]
approxiately like this: https://reactnative.dev/docs/running-on-device

So I want to try running the application at the development stage directly on Android using USB

**Reply by wschume** [reply]
Alternatively, you could try to run the app on a virtual device on your machine? Would this work for you? 
I know that this is possible with flutter. Maybe someone with more experience with the dart/flutter interface can shed some light on that.

**Reply by nexblu** [reply]
It's a shame that my laptop isn't capable of using a virtual machine, I'm curious if it can be done directly on a real device (which uses USB) ?

**Reply by ndonkoHenri** [reply]
You can preview your app using the Flet mobile application - search for "Flet" in the app/playstore.
- https://flet.dev/docs/getting-started/testing-on-android
- https://flet.dev/docs/getting-started/testing-on-ios

**Reply by nexblu** [reply]
okay, thanks sir, I'll try

**Reply by nexblu** [reply]
![image](https://github.com/flet-dev/flet/assets/141644469/42c12782-66e4-4d65-b227-76c1e1c6254b)
sorry sir, what causes this to be a connection error? I have tried on the same network

**Reply by nexblu** [reply]
Sorry sir, I want to ask, is it possible to live debug using ADB devices?

**Comment by auphof** [comment]
I have a similar requirement, I initially tested using the  Flet mobile application, "Flet" from the app/playstore. but ran into issues with  http requests locally against a local rest api in simultaneous development and I have additional Native python modules added into the P4A distro, and I think i had issues with Audio player and Audio recorder.  I have a custom `p4a` based build, mostly for additional Native python modules created through recipes but also to relax Android permissions so I can make the http requests with the mobile app, the additional modules , and the audio works great.

I do 9x% of testing with a Linux app using hot reload then `flet build APK ...` with custom template and env `SERIOUS_PYTHON_P4A_DIST` set to my `p4a` distro, 

This works great but is slow build 7min + install. A pain for quick iterative experiment.  

So i thought the same can we either debug on ADB usb tethered device or build and deploy just the changes and not go through the full APK build. I am aware that @FeodorFitsner has the `flet build apk....`  https://flet.dev/docs/publish/android  flagged as WORK IN PROGRESS and `No1 in backlog`. Would be great if there was a pathway to do hot reload on the mobile device and debugging `debugpy`   from VSCODE. As is this `flet`  ui coding with `flutter` and near universal deployment is pretty bloody great 🚀, but there are challenges and anything that can make that iterative development on target faster would help. Thanks all
